neps/nep-0509.md

```diff
@@ -0,0 +1,231 @@
+
+We also propose a change to the validator roles and responsibilities. This is the list of roles after the proposal, with same and new behavior clearly labelled:
+
+* Block producers:
+  * (Same as today) Produce blocks, (new) including waiting for chunk endorsements
+  * (Same as today) Maintain chunk parts (i.e. participates in data availability based on Reed-Solomon erasure encoding)
+  * (Same as today) Do not require tracking any shard
+  * (Same as today) Should have a higher barrier of entry for security reasons (e.g. to make block double signing harder)
+* Chunk producers:
+  * (Same as today) Produce chunks, (new) including producing chunk state witnesses
+  * (New) Distributes state witnesses to chunk validators
+  * (Same as today) Must track the shard it produces the chunk for
+  * (Same as today) Rotate shards across epoch boundaries, (new) but at a lower rate (e.g. 1 week)
+* Block validators:
+  * (Same as today) Validate blocks, (new) including verifying chunk endorsements
+  * (Same as today) Vote for blocks with endorsement or skip messages
+  * (New) No longer require tracking any shard
+  * (Same as today) Must collectively have a majority of all the validator stake, for security reasons.
+* (New) Chunk validators:
+  * Validate state witnesses, and sends chunk endorsements to block producers
+  * Do not require tracking any shard
+  * Must collectively have a majority of all the validator stake, to ensure the security of chunk validation.
+
+See the Validator Role Change section for more details.
+
+## Chunk Validator Shuffling
+
+Chunk validators will be randomly assigned to validate shards, for each block (or as we may decide later, for multiple blocks in a row, if required for performance reasons). A chunk validator may be assigned multiple shards at once, if it has sufficient stake.
+
+Each chunk validator's stake is divided into "mandates". There are full and partial mandates. The amount of stake for a full mandate is a fixed parameter determined by the stake distribution of all validators, and any remaining amount smaller than a full mandate is a partial mandate. A chunk validator therefore has zero or more full mandates plus up to one partial mandate. The list of full mandates and the list of partial mandates are then separately shuffled and partitioned equally (as in, no more than one mandate in difference between any two shards) across the shards. Any mandate assigned to a shard means that the chunk validator who owns the mandate is assigned to validate that shard. Because a chunk validator may have multiple mandates, it may be assigned multiple shards to validate.
+
+We have done research to show that the security of this algorithm is sufficient with a reasonable number of chunk validators and a reasonable number of shards, assuming a reasonable bound for the total stake of malicious nodes. TODO: Include or link to that research here.
+
```

Comment:
I think it's worth having a separate section for followings:
* state witness compression + size limit
* state witness breakdown with reed solomon encoding
* network optimization needed for each node
* reward calculation change

cc. @pugachAG , @Longarithm , @saketh-are , @shreyan-gupta , @jancionear